### PR TITLE
[wptserve] Make cookies always binary strings

### PR DESCRIPTION
**Pull request by Hexcles**
This is built on top of #23284 and I'm happy to combine the two PRs if people prefer that way.

Similar to #23284 , there'll be lots of regressions in handlers that use `request.cookies` in Python 3.

### CODE REVIEW

**Comment by Hexcles** [comment]
cc @stephenmcgruer

File: tools/wptserve/tests/functional/test_cookies.py [hunk 1]
```diff
@@ -47,18 +47,20 @@ def handler(request, response):
 
         self.assertEqual(parts["name"], "")
```
Comment (annevk) [inline tools/wptserve/tests/functional/test_cookies.py:48]: Wouldn't these have to change as well?

File: tools/wptserve/wptserve/request.py [hunk 2]
```diff
@@ -610,8 +607,33 @@ def from_field_storage(cls, fs):
...
+
+    def value_decode(self, val):
+        """Decode value from network to (real_value, coded_value).
+
+        Override BaseCookie.value_decode."""
+        real = val.encode('iso-8859-1') if isinstance(val, text_type) else val
+        return real, val
+
+    def value_encode(self, val):
+        raise NotImplementedError('BinaryCookieParser is not for setting cookies')
```
Comment (annevk) [inline tools/wptserve/wptserve/request.py:623]: Why not? How would you set a value to 0x80?

Comment (Hexcles) [reply tools/wptserve/wptserve/request.py:623]: This is only used in `Request`, and only as a parser. First, `request.cookies` are not intended to be changed (you'd want to change `response.cookies` instead); second, even if you do, `request.cookies` is the `Cookies` class not this class (this parser class gets discarded immediately).

File: tools/wptserve/tests/functional/test_cookies.py (hunk 1 above)
Comment (Hexcles) [reply tools/wptserve/tests/functional/test_cookies.py:48]: No, `resp` is actually the return value from `urllib.open` and is only used in these functional tests.

File: tools/wptserve/wptserve/request.py [hunk 3]
```diff
@@ -610,8 +607,33 @@ def from_field_storage(cls, fs):
...
+class BinaryCookieParser(BaseCookie):
+    """A subclass of BaseCookie that returns values in binary strings
+
+    This is not intended to store the cookies; use Cookies instead."""
+
+    def value_decode(self, val):
+        """Decode value from network to (real_value, coded_value).
+
+        Override BaseCookie.value_decode."""
+        real = val.encode('iso-8859-1') if isinstance(val, text_type) else val
```
Comment (jgraham) [inline tools/wptserve/wptserve/request.py:619]: Can we use `isomorphic_encode` from the other PR?

File: tools/wptserve/wptserve/response.py [hunk 4]
```diff
@@ -154,6 +159,8 @@ def maybe_set(key, value):
 
     def unset_cookie(self, name):
         """Remove a cookie from those that are being sent with the response"""
+        if PY3:
+            name = name.decode('iso-8859-1')
```
Comment (jgraham) [inline tools/wptserve/wptserve/response.py:163]: How do we know this is a type that can be decoded? Should we be using `isomorphic_decode`?

File: tools/wptserve/wptserve/response.py [hunk 5]
```diff
@@ -113,14 +113,19 @@ def set_cookie(self, name, value, path="/", domain=None, max_age=None,
                         time or interval from now when the cookie expires
 
         """
+        if value is None:
+            value = b''
+            max_age = 0
+            expires = timedelta(days=-1)
+
+        if PY3:
+            name = name.decode('iso-8859-1')
```
Comment (jgraham) [inline tools/wptserve/wptserve/response.py:122]: `isomorphic_decode`?

File: tools/wptserve/wptserve/request.py (hunk 3 above)
Comment (Hexcles) [reply tools/wptserve/wptserve/request.py]: Done. I hadn't rebased this branch.

File: tools/wptserve/wptserve/response.py (hunk 4 above)
Comment (Hexcles) [reply tools/wptserve/wptserve/response.py:165]: Well, we expect this API to only accept bytes, which will be enforced once we require all string literals to be prefixed, but yeah I've switched to `isomorphic_decode` which can accept text strings, too.

**Review by stephenmcgruer — APPROVED** [review]

File: tools/wptserve/wptserve/request.py [hunk 6]
```diff
@@ -590,8 +588,32 @@ def from_field_storage(cls, fs):
...
+        return isomorphic_encode(val), val
+
+    def value_encode(self, val):
+        raise NotImplementedError('BinaryCookieParser is not for setting cookies')
+
+    def load(self, rawdata):
+        assert isinstance(rawdata, binary_type)
+        if PY3:
+            rawdata = isomorphic_decode(rawdata)
+        # BaseCookie.load expects a native string.
```
Comment (Hexcles) [inline tools/wptserve/wptserve/request.py:609]:
```suggestion
            # BaseCookie.load expects a native string, which in Python 3 is text.
            rawdata = isomorphic_decode(rawdata)
```

File: tools/wptserve/wptserve/response.py [hunk 7]
```diff
@@ -115,14 +115,19 @@ def set_cookie(self, name, value, path="/", domain=None, max_age=None,
                         time or interval from now when the cookie expires
 
         """
```
Comment (Hexcles) [inline tools/wptserve/wptserve/response.py:117]:
```suggestion
        """
        # TODO: Convert other parameters (e.g. path) to bytes, too.
```

**Comment by Hexcles** [comment]
Rebased this and I'll merge it after the tests pass.

Fun fact: Sphinx complained about the docstring of `BaseCookieParser.load` which was inherited from the stdlib `BaseCookie.load` so I had to override the docstring, too (of course, with some actually useful information).

**Pull request merged**